### PR TITLE
Tendermint integration

### PR DESCRIPTION
**Pull request by pienkowb**

### CODE REVIEW

File: chain/tendermint/src/hash.rs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+            Ok(Self::from_str(&h).map_err(|e| D::Error::custom(format!("{}", e)))?)
+        }
+    }
+}
+
+impl Serialize for Hash {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        self.to_string().serialize(serializer)
+    }
+}
```
Comment (roignpar) [inline chain/tendermint/src/hash.rs:115]: @lukanus the `hash.rs` file is never used anywhere. is all this code needed?

File: chain/tendermint/src/trigger.rs [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                    self.block_number(),
+                    // TODO: Add `self.block_hash()`
+                )
+            }
+        }
+    }
+}
+
+pub struct EventData {
+    pub event: codec::Event, // REVIEW: Do we want to have this behind an `Arc` wrapper?
```
Comment (roignpar) [inline chain/tendermint/src/trigger.rs:132]: IMO `Arc` shouldn't be used by default, only if it's needed. the compiler will definitely let us know if that will be the case after future changes.

File: chain/tendermint/src/types.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct TendermintPartSetHeader {
+    pub total: u32,
+    pub hash: Hash,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct TendermintBlockTxData {
+    pub txs: Vec<Vec<u8>>,
+}
```
Comment (roignpar) [inline chain/tendermint/src/types.rs:116]: @lukanus `types.rs` is also not used anywhere. is it still needed?

File: graph/src/blockchain/types.rs [hunk 4]
```diff
@@ -148,6 +148,16 @@ impl From<(H256, i32)> for BlockPtr {
     }
 }
 
+impl From<(Vec<u8>, u64)> for BlockPtr {
+    fn from((bytes, number): (Vec<u8>, u64)) -> Self {
+        let number = i32::try_from(number).unwrap();
```
Comment (roignpar) [inline graph/src/blockchain/types.rs:153]: is `number` guaranteed to always fit in `i32`?

File: chain/tendermint/src/hash.rs (hunk 1 above)
Comment (lukanus) [reply chain/tendermint/src/hash.rs:115]: I've placed this intialy for creation of cosmos-based addresses. It the data gets to the node correctly decoded (so you have cosmos addresses in cosmos and terra in terra) you can skip that @roignpar

File: graph/src/blockchain/types.rs (hunk 4 above)
Comment (lukanus) [reply graph/src/blockchain/types.rs:153]: @roignpar for this one you need to talk to @lavabyrd  :)

File: chain/tendermint/src/types.rs (hunk 3 above)
Comment (lukanus) [reply chain/tendermint/src/types.rs:116]: @roignpar Not sure how about the latest version but it was used before :) as a node counterpart of ASC types

File: chain/tendermint/src/hash.rs (hunk 1 above)
Comment (pienkowb) [reply chain/tendermint/src/hash.rs:115]: This actually might be useful, I don't think we do any decoding on addresses right now.

File: server/index-node/src/resolver.rs [hunk 5]
```diff
@@ -247,7 +264,6 @@ where
         let mut response: BTreeMap<String, r::Value> = BTreeMap::new();
         response.insert("features".to_string(), features);
         response.insert("errors".to_string(), errors);
-        response.insert("network".to_string(), network);
```
Comment (patiee) [inline server/index-node/src/resolver.rs:250]: network is deleted from response intentionally?

File: server/index-node/src/resolver.rs [hunk 6]
```diff
@@ -103,8 +103,8 @@ where
             .expect("Valid blockNumber required")
             .try_into()
             .unwrap();
-
-        let block_hash = argument_values
+        // TODO(Tendermint): check this type
```
Comment (patiee) [inline server/index-node/src/resolver.rs:106]: block hash is correct right?

**Review by patiee — CHANGES_REQUESTED** [review]

File: server/index-node/src/resolver.rs (hunk 5 above)
Comment (roignpar) [reply server/index-node/src/resolver.rs:250]: i added this back in the origin merging/conflict resolving PR: https://github.com/figment-networks/graph-node/pull/7/files#diff-0238e3645398f2a041a40ad74876a35eff59db4b2f1a0bd6b6e795efc7997d25R267

File: chain/tendermint/src/runtime/abi.rs [hunk 7]
```diff
@@ -89,7 +89,20 @@ impl ToAscObj<AscHeader> for codec::Header {
             app_hash: asc_new(heap, &Bytes(&self.app_hash))?,
             last_results_hash: asc_new(heap, &Bytes(&self.last_results_hash))?,
             evidence_hash: asc_new(heap, &Bytes(&self.evidence_hash))?,
-            proposer_address: asc_new_or_null(heap, &self.proposer_address)?,
+            proposer_address: asc_new(heap, &Bytes(&self.proposer_address))?,
```
Comment (lavabyrd) [inline chain/tendermint/src/runtime/abi.rs:92]: did we mean to drop the null here?

File: chain/tendermint/src/runtime/abi.rs [hunk 8]
```diff
@@ -301,7 +380,7 @@ impl ToAscObj<AscCommitSig> for codec::CommitSig {
     ) -> Result<AscCommitSig, DeterministicHostError> {
         Ok(AscCommitSig {
             block_id_flag: asc_new(heap, &BlockIDKind(self.block_id_flag))?,
-            validator_address: asc_new_or_null(heap, &self.validator_address)?,
+            validator_address: asc_new(heap, &Bytes(&self.validator_address))?,
```
Comment (lavabyrd) [inline chain/tendermint/src/runtime/abi.rs:383]: again we've dropped the null. I just want to make sure its meant

File: chain/tendermint/src/types.rs [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+pub struct TendermintBlock {
+    pub hash: Hash,
+    pub number: u64,
+    pub parent_hash: Option<Hash>,
+    pub parent_number: Option<u64>,
+
+    pub header: TendermintBlockHeader,
+    pub data: TendermintBlockTxData,
+    //pub evidence:  Option<TendermintEvidenceList>,
+    //pub last_commit:  Option<TendermintCommit>,
```
[the remark this answers is not in the document]
Comment (roignpar) [reply chain/tendermint/src/types.rs]: removed the file, the entire module wasn't used

File: chain/tendermint/src/hash.rs [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+impl Default for Hash {
+    fn default() -> Self {
+        Hash([0u8; 32])
+    }
+}
+
+impl Display for Hash {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let hex = Hex::upper_case().encode_to_string(self.0).unwrap();
+        //Hash::Sha256(ref h) =>
```
[the remark this answers is not in the document]
Comment (roignpar) [reply chain/tendermint/src/hash.rs]: removed this file too, it wasn't used anywhere

File: server/index-node/src/resolver.rs (hunk 6 above)
Comment (roignpar) [reply server/index-node/src/resolver.rs]: not sure, what do you mean by "correct"?

File: chain/tendermint/src/runtime/abi.rs (hunk 7 above)
Comment (roignpar) [reply chain/tendermint/src/runtime/abi.rs:92]: `proposer_address` and `validator_address` both went from being `Option<Address>` to `Vec<u8>` as part of the proto definition changes @patiee made.

null was used when the value was `None`, but that can't happen anymore. the `Vec` can be empty, of course.

File: chain/tendermint/src/protobuf/gogoproto.rs [hunk 11]
```diff
@@ -0,0 +1 @@
+
```
Comment (lavabyrd) [inline chain/tendermint/src/protobuf/gogoproto.rs:1]: I wonder do we still need these files btw? They seem to be empy

File: chain/tendermint/src/protobuf/google.protobuf.rs [hunk 12]
```diff
@@ -0,0 +1 @@
+
```
Comment (lavabyrd) [inline chain/tendermint/src/protobuf/google.protobuf.rs:1]: I wonder do we still need these files btw? They seem to be empy

**Comment by lavabyrd** [comment]
Can we merge this or close it perhaps?

**Pull request closed without merging**